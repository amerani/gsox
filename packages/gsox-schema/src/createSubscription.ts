import "reflect-metadata";
import { FIELD_SYMBOL, TYPE_SYMBOL } from "./constants";

function createSubscription(T): string {
      const obj = new T();
      const type = Reflect.get(obj, TYPE_SYMBOL);
      let schema = `subscription {\n\t${type} {\n`;
      const metadata = Reflect.getMetadata(FIELD_SYMBOL, obj);
      metadata.forEach((field) => {
            schema += `\t\t${field.name}\n`;
      });
      schema += `\t}\n}`;
      return schema;
}

export {
      createSubscription,
};
