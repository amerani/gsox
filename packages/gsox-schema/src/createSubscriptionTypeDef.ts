import { TYPE_SYMBOL } from "./constants";

function createSubscriptionType(types) {
      const typeNames = types
            .map((type) => new type())
            .map((instance) => Reflect.get(instance, TYPE_SYMBOL));

      let result = `type Subscription {\n`;
      typeNames.forEach((typeName) => {
            result = result.concat(`${typeName}:${typeName}\n`);
      });
      result = result.concat("}");
      return result;
}

export { createSubscriptionType };
