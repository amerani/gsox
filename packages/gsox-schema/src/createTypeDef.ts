import "reflect-metadata";
import { FIELD_SYMBOL, TYPE_SYMBOL } from "@gsox/core";

function createTypeDef(T): string {
      const obj = new T();
      const type = Reflect.get(obj, TYPE_SYMBOL);
      let schema = `type ${type} {\n`;
      const metadata = Reflect.getMetadata(FIELD_SYMBOL, obj);
      metadata.forEach((field) => {
            schema += `\t${field.name}: ${field.type}\n`;
      });
      schema += `}`;
      return schema;
}

function createTypeDefs(types): string {
      return types.reduce(
            (acc, cur) => {
                  const typeStr = createTypeDef(cur);
                  return acc.concat([typeStr, "\n"]);
            },
            "");
}

export {
      createTypeDef,
      createTypeDefs,
};
