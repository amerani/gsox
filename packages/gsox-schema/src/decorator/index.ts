export * from "./ListField";
export * from "./Type";