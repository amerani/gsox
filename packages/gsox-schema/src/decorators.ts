import { FIELD_SYMBOL, TYPE_SYMBOL } from "@gsox/core";
import "reflect-metadata";

function Type(value?) {
      return function classDecorator<T extends {new(...args: any[]): {}}>(constructor: T) {
            return class extends constructor {
                  public [TYPE_SYMBOL] = value || constructor.name;
            };
        };
}

function Field(type?, name?) {
      return function(obj, prop) {
            const id = prop;
            name = name || prop;
            type = type || Reflect.getMetadata("design:type", obj, prop).name;
            const entry = {id, name, type};
            if (Reflect.hasMetadata(FIELD_SYMBOL, obj)) {
                  const cur = Reflect.getMetadata(FIELD_SYMBOL, obj);
                  Reflect.defineMetadata(FIELD_SYMBOL, [entry, ...cur], obj);
            } else {
                  Reflect.defineMetadata(FIELD_SYMBOL, [entry], obj);
            }
      };
}

export {
      Type,
      Field,
};
