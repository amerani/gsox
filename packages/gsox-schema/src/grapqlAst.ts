import {
      DocumentNode,
      FieldNode,
      NameNode,
      OperationDefinitionNode,
      SelectionNode,
      SelectionSetNode,
    } from "graphql";

class SubscriptionDocument implements DocumentNode {
      public kind: "Document";
      public definitions: OperationDefinitionNode[];
      constructor(definitions) {
            this.definitions = definitions;
      }
}

class OperationNode implements OperationDefinitionNode {
      public kind: "OperationDefinition";
      public operation: "subscription";
      public selectionSet: SelectionSetNode;
      constructor(selectionSet) {
            this.selectionSet = selectionSet;
      }
}

class SelectionSet implements SelectionSetNode {
      public kind: "SelectionSet";
      public selections: FieldNode[];
      constructor(selections: FieldNode[]) {
            this.selections = selections;
      }
}

class NameNodeImp implements NameNode {
      public kind: "Name";
      public value: string;
      constructor(value) {
            this.value = value;
      }
}

class Field implements FieldNode {
      public kind: "Field";
      public name: NameNode;
      public selectionSet?: SelectionSetNode;
      constructor(name, selectionSet?) {
            this.name = new NameNodeImp(name);
            this.selectionSet = selectionSet;
      }
}

export { Field, SelectionSet, OperationNode, SubscriptionDocument };
