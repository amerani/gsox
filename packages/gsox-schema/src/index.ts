export * from "./decorators";
export * from "./typeDef";
export * from "./subscription";
export * from "./typeSub";