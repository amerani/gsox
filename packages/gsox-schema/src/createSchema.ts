import gql from "graphql-tag";
import "reflect-metadata";
import { createSubscriptionType } from "./createSubscriptionTypeDef";
import { createTypeDefs } from "./createTypeDef";

const rootSchema = `
      type Query {
            hello: String
      }
      type schema {
            subscription: Subscription
            query: Query
      }
      type Ping {
            id: Int
      }
      type Subscription {
            Ping: Ping
      }
`;
function createTypeDefString(types) {
      return `
            ${createSubscriptionType(types)} \n
            ${createTypeDefs(types)}
      `;
}

function createSchemaString(types) {
      return `
            ${rootSchema} \n
            ${createTypeDefString(types)}
      `;
}

function createSchema(types) {
      return gql(createSchemaString(types));
}

export {
      createSchemaString,
      createSchema,
};
