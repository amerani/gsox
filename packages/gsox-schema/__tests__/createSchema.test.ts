import { Field, Type } from "../src";
import { createSchema, createSchemaString } from "../src/createSchema";

@Type()
class Ping {
      @Field("Int")
      public id: number;
}

@Type()
class Message {
      @Field()
      public content: string;
}
test("should output schema", () => {
      const res = createSchemaString([Ping, Message]);
      expect(res).toMatchSnapshot();
});

test("should build schema", () => {
      const res = createSchema([Ping, Message]);
      expect(res).toMatchSnapshot();
});
