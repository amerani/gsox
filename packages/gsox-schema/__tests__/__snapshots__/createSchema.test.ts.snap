// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should build schema 1`] = `
Object {
  "definitions": Array [
    Object {
      "description": undefined,
      "directives": Array [],
      "fields": Array [
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "hello",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "String",
            },
          },
        },
      ],
      "interfaces": Array [],
      "kind": "ObjectTypeDefinition",
      "name": Object {
        "kind": "Name",
        "value": "Query",
      },
    },
    Object {
      "description": undefined,
      "directives": Array [],
      "fields": Array [
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "subscription",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "Subscription",
            },
          },
        },
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "query",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "Query",
            },
          },
        },
      ],
      "interfaces": Array [],
      "kind": "ObjectTypeDefinition",
      "name": Object {
        "kind": "Name",
        "value": "schema",
      },
    },
    Object {
      "description": undefined,
      "directives": Array [],
      "fields": Array [
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "id",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "Int",
            },
          },
        },
      ],
      "interfaces": Array [],
      "kind": "ObjectTypeDefinition",
      "name": Object {
        "kind": "Name",
        "value": "Ping",
      },
    },
    Object {
      "description": undefined,
      "directives": Array [],
      "fields": Array [
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "Ping",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "Ping",
            },
          },
        },
      ],
      "interfaces": Array [],
      "kind": "ObjectTypeDefinition",
      "name": Object {
        "kind": "Name",
        "value": "Subscription",
      },
    },
    Object {
      "directives": Array [],
      "fields": Array [
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "Ping",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "Ping",
            },
          },
        },
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "Message",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "Message",
            },
          },
        },
      ],
      "interfaces": Array [],
      "kind": "ObjectTypeExtension",
      "name": Object {
        "kind": "Name",
        "value": "Subscription",
      },
    },
    Object {
      "description": undefined,
      "directives": Array [],
      "fields": Array [
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "id",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "Int",
            },
          },
        },
      ],
      "interfaces": Array [],
      "kind": "ObjectTypeDefinition",
      "name": Object {
        "kind": "Name",
        "value": "Ping",
      },
    },
    Object {
      "description": undefined,
      "directives": Array [],
      "fields": Array [
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "content",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "String",
            },
          },
        },
      ],
      "interfaces": Array [],
      "kind": "ObjectTypeDefinition",
      "name": Object {
        "kind": "Name",
        "value": "Message",
      },
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 432,
    "start": 0,
  },
}
`;

exports[`should output schema 1`] = `
"
            
      type Query {
            hello: String
      }
      type schema {
            subscription: Subscription
            query: Query
      }
      type Ping {
            id: Int
      }
      type Subscription {
            Ping: Ping
      }
 

            
            extend type Subscription {
Ping:Ping
Message:Message
} 

            type Ping {
	id: Int
},
type Message {
	content: String
},

      
      "
`;
